File: 2020 카카오 인턴십/경주로 건설.cpp
#include <iostream>
#include <string>
#include <cstring>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

const int MAX_N = 25, INF = 987654321;

struct Car {
    int y, x;
    int dir;
    int cost;
    Car(int _y, int _x, int _dir, int _cost): y(_y), x(_x), dir(_dir), cost(_cost) {}    
};

int n, answer = INF;
queue<Car> q;
int discovered[MAX_N][MAX_N][4];

int dy[] = {-1,1,0,0};
int dx[] = {0,0,-1,1};

bool checkVertical(int d, int nd) {
    if(d == 0 && nd == 2) return true;
    if(d == 0 && nd == 3) return true;
    if(d == 1 && nd == 2) return true;
    if(d == 1 && nd == 3) return true;
    if(d == 2 && nd == 0) return true;
    if(d == 2 && nd == 1) return true;
    if(d == 3 && nd == 0) return true;
    if(d == 3 && nd == 1) return true;
    return false;
}

void bfs(const vector<vector<int> >& board) {
    memset(discovered, -1, sizeof(discovered));
    q.push(Car(0, 0, 1, 0));
    q.push(Car(0, 0, 3, 0));
    while(!q.empty()) {
        Car now = q.front();
        q.pop();
        if(now.y == (n-1) && now.x == (n-1)) {
            answer = min(answer, now.cost);
            continue;
        }
        for(int i = 0; i < 4; ++i) {
            int ny = now.y + dy[i];
            int nx = now.x + dx[i];
            int newAddCost = 100;
            if(checkVertical(now.dir, i))
                newAddCost += 500;
            // board 안에 위치하고
            if(ny >= 0 && nx >= 0 && ny < n && nx < n) {
                // 벽이 아니고
                if(board[ny][nx] != 1) {
                    // 만약 방문하지 않았거나, 이전에 방문했었을 때의 경주로 비용보다 싸다면
                    if(discovered[ny][nx][i] == -1 || (discovered[ny][nx][i] > now.cost + newAddCost)) {
                        discovered[ny][nx][i] = now.cost + newAddCost;
                        q.push(Car(ny, nx, i, discovered[ny][nx][i]));
                    }
                }
            }
        }
    }
}

int solution(vector<vector<int>> board) {
    n = board.size();
    bfs(board);
    return answer;
}

File: 2020 카카오 인턴십/경주로 건설.java
import java.util.*;

class Solution {
    class Car {
        Car(int y, int x, int d, int cost) {
            this.y = y;
            this.x = x;
            this.d = d;
            this.cost = cost;
        }
        int y, x;
        int d;
        int cost;
    }
    
    int N, answer;
    Queue<Car> myQue = new LinkedList<>();
    boolean[][][] discovered = new boolean[N][N][4];
    int[][] D = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
    
    public int solution(int[][] board) {
        N = board.length;
        
        myQue.add(new Car(0, 0, 1, 0));
        myQue.add(new Car(0, 0, 3, 0));
        discovered[0][0][0] = true;
        discovered[0][0][2] = true;
        
        while(!myQue.isEmpty()) {
            Car car = myQue.poll();
            if(car.y == N-1 && car.x == N-1) {
                answer = car.cost;
                break;
            }
			for(int i = 0; i < 4; ++i) {
                int ny = car.y + D[i][0];
                int nx = car.x + D[i][1];
                if(ny < 0 || ny >= N || nx < 0 || nx >= N)
                    continue;
                if(board[ny][nx] != 1 && !discovered[ny][nx][i]) {
                    discovered[ny][nx][i] = true;
                    if((i / 2) != (car.d / 2)) 
                        myQue.add(new Car(ny, nx, i, car.cost + 600));
                    else
                        myQue.add(new Car(ny, nx, i, car.cost + 100));
                }
            }            
        }
        
        return answer;
    }
}
